62.CAD/main.cpp

```diff
@@ -845,18 +1877,45 @@ struct DrawBuffersFiller
 		currentObjectInSection += objectsToUpload;
 	}
 
-	// TODO[Lucas] addHatch_Internal with similar signature to functions above. 
-	/*
-	* this does:
-		- iterates through the "Boxes" in a hatch
-		- finds the curves referenced by it and copies both the boxes and the curves into correct places of the geometry buffer
-		- constructs drawObjs and copies them into correct place of the drawobj buffer
-		- and correctly advances the memory tracker and counters
-		- it will iterate to a point where it ends OR there is not enough memory left
-		- For example we might have enough memory left for 10 curve boxes in the geometry buffer,
-			but that doesn't mean we could draw 10 curve boxes because their curves need to also reside in mem,
-			the solution is simple when we iterate on curve boxes and keep track of what curves we have already copied into mem (a map from cpuCurveIndex to geomBufferOffset)
-	*/
+	void addHatch_Internal(const Hatch& hatch, uint32_t& currentObjectInSection, uint32_t mainObjIndex)
+	{
+		constexpr uint32_t IndicesPerHatchBox = 6u;
+		uint32_t uploadableObjects = (maxIndices - currentIndexCount) / IndicesPerHatchBox;
+		uploadableObjects = core::min(uploadableObjects, maxDrawObjects - currentDrawObjectCount);
+		uploadableObjects = core::min(uploadableObjects, maxGeometryBufferSize - currentGeometryBufferSize);
+
+		uint32_t i = 0;
+		for (; i + currentObjectInSection < hatch.hatchBoxes.size() && i < uploadableObjects; i++)
```

Comment:
instead of doing this weird for loop, be consistent with what other similar functions do:
```
uint32_t remainingHatches = hatch.hatchBoxes.size() - currentObjectInSection;
uint32_t objectsToUpload = core::min(uploadableObjects, remainingHatches);

for (uint32_t i = 0u; i < objectsToUpload; ++i)
{
...
}

currentDrawObjectCount += objectsToUpload;
currentObjectInSection += objectsToUpload;
```